connectors/sources/gcs.py

```diff
@@ -0,0 +1,336 @@
+    "content_encoding": "contentEncoding",
+    "content_language": "contentLanguage",
+    "created_at": "timeCreated",
+    "last_updated": "updated",
+    "metadata": "metadata",
+    "name": "name",
+    "size": "size",
+    "storage_class": "storageClass",
+    "_timestamp": "updated",
+    "type": "contentType",
+    "url": "selfLink",
+    "version": "generation",
+    "bucket_name": "bucket",
+}
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_WAIT_MULTIPLIER = 2
+DEFAULT_CONTENT_EXTRACTION = True
+DEFAULT_FILE_SIZE_LIMIT = 10485760
+STORAGE_EMULATOR_HOST = os.getenv(
+    key="STORAGE_EMULATOR_HOST", default=None
+)  # For end to end tests set this environment variable with target host.
+RUNNING_FTEST = bool(
+    os.getenv(key="RUNNING_FTEST", default=None)
+)  # Flag to check if a connector is run for ftest or not.
+DEFAULT_PEM_FILE = (
+    os.path.join(
+        os.path.dirname(__file__),
+        "tests/fixtures/gcs",
+        "service_account_dummy_cert.pem",
+    ),
+)[0]
+
+
+class GoogleCloudStorageDataSource(BaseDataSource):
+    """Class to fetch documents from Google Cloud Storage."""
+
+    def __init__(self, connector):
+        """Setup connection to the Google Cloud Storage Client.
+
+        Args:
+            connector (BYOConnector): Object of the BYOConnector class.
+        """
+        super().__init__(connector=connector)
+        if not self.configuration["service_account_credentials"]:
+            raise Exception("service_account_credentials can't be empty.")
+
+        self.credentials = (
+            self.configuration["service_account_credentials"]
+            .strip()
+            .encode("unicode_escape")
+            .decode()
+        )
+        self.service_account_credentials = ServiceAccountCreds(
+            scopes=[CLOUD_STORAGE_READ_ONLY_SCOPE],
+            **json.loads(
+                self.configuration["service_account_credentials"]
+                if RUNNING_FTEST
+                else self.credentials
+            ),
+        )
+        self.user_project_id = self.service_account_credentials.project_id
+        self.retry_count = int(
+            self.configuration.get("retry_count", DEFAULT_RETRY_COUNT)
+        )
+        self.enable_content_extraction = self.configuration.get(
+            "enable_content_extraction", DEFAULT_CONTENT_EXTRACTION
+        )
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for Google Cloud Storage.
+
+        Returns:
+            dictionary: Default configuration.
+        """
+        return {
+            "service_account_credentials": {
+                "value": f'{{"type": "service_account","project_id": "dummy_project_id","private_key_id": "abc","private_key": {json.dumps(open(DEFAULT_PEM_FILE).read())} ,"client_email": "123-abc@developer.gserviceaccount.com","client_id": "123-abc.apps.googleusercontent.com","auth_uri": "https://accounts.google.com/o/oauth2/auth","token_uri": "http://localhost:443/token"}}',
```

Comment:
to make this error-proof and more readable, you can prepare a JSON mapping in a separate variable:

```
credentials = {'type': ...
}
return {
            "service_account_credentials": {
                "value":  json.dumps(crdentials)
...
```